client/services/LocalStorageService.js

```diff
@@ -0,0 +1,122 @@
+// Copyright 2017 The TIE Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/*
+ * A service that saves student code and feedback to the browser's localStorage.
+ */
+tie.factory('LocalStorageService', ['FeedbackParagraphObjectFactory',
+  function(FeedbackParagraphObjectFactory) {
+    // In some browsers, localStorage is not available and its
+    // invocation throws an error.
+    var localStorageIsAvailable = false;
+    try {
+      localStorageIsAvailable = Boolean(localStorage);
+    } catch (e) {
+      localStorageIsAvailable = false;
+    }
+
+    var getLocalStorageKeyForCode = function(questionId, language) {
+      return questionId + ":" + language;
+    };
+
+    var getLocalStorageKeyForFeedback = function(questionId, language) {
+      return questionId + ":feedback:" + language;
```

Comment:
Do we actually need the language here? If the user changes the language, does the existing feedback disappear? If not, what happens to it?

Also, the discrepancy between the key for code and the key for feedback is a bit odd. I think we may want to backtrack a bit and make a short doc that outlines the desired namespace structure for localStorage (for storing question completion, code, feedback). What conventions would we use if we were coming up with this from scratch, now that we have a better idea of all the data we need to persist?